util/topsql/reporter/reporter.go

```diff
@@ -209,14 +229,91 @@ func (tsr *RemoteTopSQLReporter) Close() {
 	tsr.client.Close()
 }
 
+func addEvictedCPUTime(collectTarget map[string]*dataPoints, timestamp uint64, totalCPUTimeMs uint32) {
+	if totalCPUTimeMs == 0 {
+		return
+	}
+	others, ok := collectTarget[keyOthers]
+	if !ok {
+		others = &dataPoints{}
+		collectTarget[keyOthers] = others
+	}
+	if len(others.TimestampList) == 0 {
+		others.TimestampList = []uint64{timestamp}
+		others.CPUTimeMsList = []uint32{totalCPUTimeMs}
+	} else {
+		others.TimestampList = append(others.TimestampList, timestamp)
+		others.CPUTimeMsList = append(others.CPUTimeMsList, totalCPUTimeMs)
+		length := len(others.TimestampList)
+		if others.TimestampList[length-1] < others.TimestampList[length-2] {
+			// This situation appears only when time jump backward.
+			// If so, sort the data points by timestamp.
+			sort.Sort(others)
+		}
+	}
+	others.CPUTimeMsTotal += uint64(totalCPUTimeMs)
+}
+
+func (d *dataPoints) merge(other *dataPoints) {
+	if other == nil || len(other.TimestampList) == 0 {
+		return
+	}
+	if other.isInvalid() {
+		logutil.BgLogger().Warn("[top-sql] data points is invalid, it should never happen", zap.Reflect("self", *d), zap.Reflect("other", *other))
+		return
+	}
+	if !sort.IsSorted(other) {
+		// This situation appears only when time jump backward.
+		// If so, sort the data points by timestamp.
+		sort.Sort(other)
+	}
+	if len(d.TimestampList) == 0 {
+		d.TimestampList = other.TimestampList
+		d.CPUTimeMsList = other.CPUTimeMsList
+		d.CPUTimeMsTotal = other.CPUTimeMsTotal
+		return
+	}
+	d.CPUTimeMsTotal += other.CPUTimeMsTotal
+	findIdx := 0
+	for idx, ts := range other.TimestampList {
+		cpuTimeMs := other.CPUTimeMsList[idx]
+		for ; findIdx < len(d.TimestampList); findIdx++ {
+			findTs := d.TimestampList[findIdx]
+			if ts == findTs {
+				d.CPUTimeMsList[findIdx] += cpuTimeMs
+				break
+			}
+			if ts < findTs {
+				d.TimestampList = append(d.TimestampList, 0)
+				copy(d.TimestampList[findIdx+1:], d.TimestampList[findIdx:])
+				d.TimestampList[findIdx] = ts
+				d.CPUTimeMsList = append(d.CPUTimeMsList, 0)
+				copy(d.CPUTimeMsList[findIdx+1:], d.CPUTimeMsList[findIdx:])
+				d.CPUTimeMsList[findIdx] = cpuTimeMs
+				break
+			}
+		}
+		if ts > d.TimestampList[len(d.TimestampList)-1] {
+			d.TimestampList = append(d.TimestampList, ts)
+			d.CPUTimeMsList = append(d.CPUTimeMsList, cpuTimeMs)
+		}
+	}
```

Comment:
This process is O(N*M) in worse case where `N = len(d.ts)` and `M = len(other.ts)`:

```
d.ts = [1, 999, 1000, 1001, 1002, ...]
other.ts = [2, 3, 4, 5, 6, ...]
```

Please optimize it.